### DISCUSSION TITLE
Batch Auto-Deskew/Rotate?

### DISCUSSION THREAD

**Discussion by Karyudo** [ImageMagick/ImageMagick#2511] [General]
I scanned a bunch of magazine pages, and the resulting images are generally close to being square/rectilinear, but also they're not *exactly* horizontal/vertical. 

Is there a way to automatically detect the skew angle and deskew the image, for each image in a batch?

**Comment by snibgo** [comment]
The "-deskew" operation does that. See https://www.imagemagick.org/script/command-line-options.php#deskew

"magick mogrify" is designed to process a wild-card set of files. See https://www.imagemagick.org/script/mogrify.php

An alternative is to use a shell loop, and call "magick" for each file.

**Reply by Karyudo** [reply]
Is there a way to process a folder of files (with '*.jpg', say) and add a suffix to the input filename for use as the output filename, so I can A/B the results?

I've tried deskewing without mogrify, which just uses a single new filename, and adds numbered suffixes, such that the new files don't sort interleaved with the originals. With mogrify, the filename is kept, but output completely replaces the input file. Neither is ideal.

Also, how can I keep -deskew from making some files considerably worse than the input? One of my half-dozen test files starts out within a few tenths of a degree of square, and the output is several degrees off.

**Comment by snibgo** [comment]
"magick mogrify" is useful when we don't want to change filenames. "-path" puts the output files in a different directory, which might be useful for A/B comparisons.

"magick" (with no "mogrify") can be used in a shell loop, so each call of "magick" processes a single input, and you can have any output filename. For example:
```
magick myfile.png -set filename:MYFILE %t_B.%e %[filename:MYFILE]
```
... will create an output named "myfile_B.png".

See https://www.imagemagick.org/script/escape.php for the escapes.

Or you can use shell facilities for filenames. For Windows, see "help for" for %~nI etc.


"-deskew" has a limited precision. I forget how much, perhaps just one degree. It won't work with more accuracy than that.
